### PR TITLE
pkg/trace/api: add probabilistic sampling for OTLP

### PR DESCRIPTION
**Pull request by gbbr**
This change introduces a new probabilistic sampler as part of the OTLP Ingest in the trace-agent. This sampler allows the user to control ingestion reliably in a distributed system, considering that the same percentage is used in all agents.

Use the following configuration:
```yaml
otlp_config:
  traces:
    probabilistic_sampler:
      # @param sampling_percentage - number - optional - default: 100
      # @env DD_OTLP_CONFIG_TRACES_PROBABILISTIC_SAMPLER_SAMPLING_PERCENTAGE - number - optional - default: 100
      # Percentage of traces to ingest (0 100]. Invalid values (<= 0 || > 100) are disconsidered
      # and the default is used. If incoming spans have a sampling.priority set by the user, it 
      # will be followed and the sampling percentage will be overridden.
      sampling_percentage: 100
```

### CODE REVIEW

**Review by kayayarai — APPROVED** [review]
👍 from docs (release note)

File: pkg/trace/api/otlp.go [hunk 1]
```diff
@@ -136,6 +137,39 @@ func (o *OTLPReceiver) processRequest(ctx context.Context, header http.Header, i
...
 
+// knuthFactor represents a large, prime number ideal for Knuth's Multiplicative Hashing.
+// Warning: do not change this number. It is shared with other probabilistic samplers
+// in the agent, the Datadog libraries, and in OpenTelemetry. This ensures consistency
+// in a distributed system.
+const knuthFactor = uint64(1111111111111111111)
+
+// samplingRate returns the rate as defined by the probabilistic sampler.
+func (o *OTLPReceiver) samplingRate() float64 {
+	rate := float64(o.conf.OTLPReceiver.ProbabilisticSampling) / 100
```
Comment (songy23) [inline pkg/trace/api/otlp.go:148]: Isn't `o.conf.OTLPReceiver.ProbabilisticSampling` already float64?

File: pkg/config/config_template.yaml [hunk 2]
```diff
@@ -3483,6 +3483,18 @@ api_key:
...
 
+    ## @param probabilistic_sampler - custom object - optional
+    ## Probabilistic sampler controlling the rate of ingestion. Using this sampler works consistently
+    ## in a distributed system where the sampling rate is shared. Exceptions are made for errors and
+    ## rare traces (if enabled via apm_config.enable_rare_sampler).
+    #
+    # probabilistic_sampler:
+      ## @param sampling_percentage - number - optional - default: 100
+      ## @env DD_OTLP_CONFIG_TRACES_PROBABILISTIC_SAMPLER_SAMPLING_PERCENTAGE - number - optional - default: 100
+      ## Percentage of traces to ingest (0 100]. A value of 0 disables sampling and everything is ingested.
```
Comment (songy23) [inline pkg/config/config_template.yaml:3494]: Probably worth mentioning what happens if sampling_percentage isn't in the valid range.

File: pkg/trace/config/config.go [hunk 3]
```diff
@@ -73,6 +73,11 @@ type OTLP struct {
 	// The 'preview' rules change the canonical hostname chosen in cloud providers to be consistent with the
 	// one sent by Datadog cloud integrations.
 	UsePreviewHostnameLogic bool `mapstructure:"-"`
+
+	// ProbabilisticSampling specifies the percentage of traces to ingest. Exceptions are made for errors
+	// and rare traces (outliers) if "RareSamplerEnabled" is true.
+	// Valid in range (0, 100].
```
Comment (songy23) [inline pkg/trace/config/config.go:79]: Same, worth mentioning what if the value is < 0 or > 100.

Comment (gbbr) [reply pkg/trace/config/config.go]: Yep. Done.

**Review by songy23 — APPROVED** [review]
LGTM

File: pkg/trace/api/otlp.go [hunk 4]
```diff
@@ -288,6 +311,34 @@ func (o *OTLPReceiver) ReceiveResourceSpans(ctx context.Context, rspans ptrace.R
...
+			Spans: spans,
+		}
+		if p, ok := prioritiesByID[k]; ok {
+			// a manual decision has been made by the user
+			chunk.Priority = int32(p)
+			traceutil.SetMeta(spans[0], "_dd.p.dm", "-4")
+		} else {
+			// we use the probabilistic sampler to decide
+			chunk.Priority = int32(o.sample(k))
+			traceutil.SetMeta(spans[0], "_dd.p.dm", "-9")
```
Comment (Barbayar) [inline pkg/trace/api/otlp.go:335]: Are we tracking `_dd.p.dm` values somewhere? How did you come up with `9`?

**Review by Barbayar — APPROVED** [review]
LGTM from intake perspective.

**Review by ahmed-mez — APPROVED** [review]
Reviewed `cmd/trace-agent/config` 👍

**Review by hush-hush — APPROVED** [review]

File: pkg/trace/api/otlp.go (hunk 4 above)
Comment (gbbr) [reply pkg/trace/api/otlp.go:335]: Answered in private. Yes.

**Pull request merged**